MediaGalleryCatalogUi/Model/Listing/DataProvider.php

```diff
@@ -0,0 +1,157 @@
+     * @param AttributeValueFactory $attributeValueFactory
+     * @param array $meta
+     * @param array $data
+     * @SuppressWarnings(PHPMD.ExcessiveParameterList)
+     */
+    public function __construct(
+        string $name,
+        string $primaryFieldName,
+        string $requestFieldName,
+        ReportingInterface $reporting,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        RequestInterface $request,
+        FilterBuilder $filterBuilder,
+        SearchResultFactory $searchResultFactory,
+        CategoryListInterface $categoryList,
+        AttributeValueFactory $attributeValueFactory,
+        array $meta = [],
+        array $data = []
+    ) {
+        parent::__construct(
+            $name,
+            $primaryFieldName,
+            $requestFieldName,
+            $reporting,
+            $searchCriteriaBuilder,
+            $request,
+            $filterBuilder,
+            $meta,
+            $data
+        );
+        $this->categoryList = $categoryList;
+        $this->searchResultFactory = $searchResultFactory;
+        $this->attributeValueFactory = $attributeValueFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getData()
+    {
+        try {
+            return $this->searchResultToOutput($this->getSearchResult());
+        } catch (\Exception $exception) {
+            return [
+                'items' => [],
+                'totalRecords' => 0,
+                'errorMessage' => $exception->getMessage()
+            ];
+        }
+    }
+    /**
+     * @inheritDoc
+     */
+    public function getSearchResult(): SearchResultInterface
+    {
+        $collection = $this->categoryList->getList($this->getSearchCriteria());
+        $items = [];
+        foreach ($collection->getItems() as $category) {
+            if ($category->getId() == 1) {
+                continue;
+            }
+            $items[] = $this->addAttributes(
+                [
+                    'name'  => $category->getName(),
+                    'entity_id' => $category->getEntityId(),
+                    'image' => $category->getImage(),
+                    'path' => $category->getPath(),
+                    'display_mode' => $category->getDisplayMode(),
+                    'products' => $category->getProductCount(),
+                    'include_in_menu' => $category->getIncludeInMenu(),
+                    'enabled' => $category->getIsActive()
+                ]
+            );
+        }
+
+        $searchResult = $this->searchResultFactory->create();
+        $searchResult->setSearchCriteria($this->getSearchCriteria());
+        $searchResult->setItems($items);
+        $searchResult->setTotalCount(count($items));
+
+        return $searchResult;
+    }
+
+    /**
+     * Add attributes to grid result
+     *
+     * @param array $attributes [code => value]
+     */
+    private function addAttributes(array $attributes)
+    {
+        $category =  new DataObject([]);
```

Comment:
I would suggest working with DocumentInterface instead of DataObject here, this interface provides the custom attribute property required and it is used as a reference inside the `Magento\Framework\Api\Search\SearchResultInterface`